src/AppCommon/Commands/ServiceControlCommand.cs

```diff
@@ -240,35 +241,37 @@ protected override async Task<EnvironmentDetails> GetEnvironment(CancellationTok
     async Task<ServiceControlEndpoint[]> GetKnownEndpoints(CancellationToken cancellationToken)
     {
         // Tool can't proceed without this data, try 5 times
-        var arr = await primary.GetData<JArray>("/endpoints", 5, cancellationToken);
+        var arr = await primary.GetData<JsonArray>("/endpoints", 5, cancellationToken);
 
-        var endpoints = arr.Select(endpointToken => new
-        {
-            Name = endpointToken["name"].Value<string>(),
-            HeartbeatsEnabled = endpointToken["monitored"].Value<bool>(),
-            ReceivingHeartbeats = endpointToken["heartbeat_information"]["reported_status"].Value<string>() == "beating"
-        })
-        .GroupBy(x => x.Name)
-        .Select(g => new ServiceControlEndpoint
+        var endpoints = arr?.Select(endpoint => new
         {
-            Name = g.Key,
-            HeartbeatsEnabled = g.Any(e => e.HeartbeatsEnabled),
-            ReceivingHeartbeats = g.Any(e => e.ReceivingHeartbeats)
+            Name = endpoint?.AsObject().TryGetPropertyValue("name", out var name) == true ? name?.GetValue<string>() : "",
+            HeartbeatsEnabled = endpoint?.AsObject().TryGetPropertyValue("monitored", out var monitored) == true ? monitored?.GetValue<bool>() : false,
+            ReceivingHeartbeats = endpoint?.AsObject().TryGetPropertyValue("heartbeat_information", out var heartbeats) == true && heartbeats?.AsObject().TryGetPropertyValue("reported_status", out var reportStatus) == true && reportStatus!.GetValue<string>() == "beating",
         })
-        .ToArray();
+            .GroupBy(x => x.Name)
+            .Select(g => new ServiceControlEndpoint
+            {
+                Name = g.Key!,
+                HeartbeatsEnabled = g.Any(e => e.HeartbeatsEnabled == true),
+                ReceivingHeartbeats = g.Any(e => e.ReceivingHeartbeats)
+            })
+            .ToArray();
 
         var useAuditCounts = false;
 
         if (primary.Version.Version >= MinAuditCountsVersion)
         {
             // Verify audit instances also have audit counts
-            var remotesInfoJson = await primary.GetData<JArray>("/configuration/remotes", cancellationToken);
+            var remotesInfoJson = await primary.GetData<JsonArray>("/configuration/remotes", cancellationToken);
```

Comment:
Part of the deserialization is already abstracted away in https://github.com/Particular/NServiceBus.ServicePlatform.Connector/blob/main/src/NServiceBus.ServicePlatform.Connector/ServicePlatformConnectionConfiguration.cs#L59-L64 but I guess we did not take a dependency to that package because it brings lots of baggage https://github.com/Particular/NServiceBus.ServicePlatform.Connector/blob/main/src/NServiceBus.ServicePlatform.Connector/NServiceBus.ServicePlatform.Connector.csproj#L15-L20

Reply:
Is it worth doing all this gymnastic vs declaring local types and simply deserialize it to the more strongly typed versions? I mean this version might have more flexibility to deal with various behaviors of ServiceControl but even a local contract would not necessarily hinder anything if we evolve it properly

Reply:
Most likely it would make things easier. At this point I just wanted to get rid of Newtonsoft. We don't know what work may be coming out of the "throughput collector in platform" task force so I want to avoid making too many changes at this point.

Reply:
Quite a bit uglier than Newtonsoft API isn't it. 😬 Here it would make sense to use an object model since we control the other end of it, though I agree that wouldn't have to be done in this PR.

Just a warning that in the Rabbit query code you would NOT want to do that because different versions of Rabbit are already known to return differently - at some version they introduced paging by completely breaking the response structure.

I could also see creating an extension method enabling us to replace Newtonsoft's `endpointToken["heartbeat_information"]["reported_status"].Value<string>()` with `endpoint.GetNestedValue<string>("heartbeat_information", "reported_status")` that took `params string[]` just for readability.